### ISSUE TITLE
[BUG] OIIO no longer builds against split IlmBase and OpenEXR builds

### ISSUE THREAD

**Issue by nrusch** [OpenImageIO/oiio#2928]
Historically we've always released IlmBase and OpenEXR to separate install trees. However, it seems that the change to support OpenEXR 3 (ab3f04204ba228ca778044c7a4e84d157695e4aa) has broken this.

I've tested trying to build 2.2.13 against split releases of OpenEXR 2.3, 2.4, and 2.5, all with the same (failing) result. I've also confirmed that OIIO *will* build against a combined IlmBase/OpenEXR 2.5 install.

Here is the CMake command I'm using in both cases:
```
    cmake
    -G "Unix Makefiles"
    -D CMAKE_INSTALL_PREFIX=${REZ_INSTALL_PREFIX}
    -D CMAKE_BUILD_TYPE=${REZ_RELEASE_TYPE}
    -D CMAKE_EXE_LINKER_FLAGS=-Wl,--enable-new-dtags
    -D CMAKE_SHARED_LINKER_FLAGS=-Wl,--enable-new-dtags
    -D CMAKE_CXX_STANDARD=11
    -D BUILD_SHARED_LIBS=ON
    -D LINKSTATIC=ON
    -D EMBEDPLUGINS=ON
    -D GLIBCXX_USE_CXX11_ABI=0
    -D USE_SIMD=sse4.2
    -D TIFF_INCLUDE_DIR=${tiff_INCLUDE_DIRS}
    -D TIFF_LIBRARY=${tiff_libtiff_STATIC_LIBRARY}
    -D PNG_PNG_INCLUDE_DIR=${png_INCLUDE_DIRS}
    -D PNG_LIBRARY=${png_libpng15_STATIC_LIBRARY}
    -D JPEG_INCLUDE_DIR=${jpeg_INCLUDE_DIRS}
    -D JPEG_LIBRARY=${jpeg_libjpeg_STATIC_LIBRARY}
    -D CMAKE_LIBRARY_PATH=${lcms_LIBRARY_DIRS}
    -D BUILD_MISSING_FMT=OFF
    -D BUILD_MISSING_PYBIND11=OFF
    -D Pybind11_ROOT=${REZ_PYBIND11_ROOT}
    -D BUILD_MISSING_ROBINMAP=OFF
    -D Robinmap_ROOT=${REZ_ROBIN_MAP_ROOT}
    -D USE_LIBRAW=ON
    -D LibRaw_ROOT=${REZ_LIBRAW_ROOT}
    -D USE_PYTHON=ON
    -D Python_ADDITIONAL_VERSIONS=${PYTHON_MAJOR_VERSION}.${PYTHON_MINOR_VERSION}
    -D USE_OPENCOLORIO=OFF
    -D USE_DCMTK=OFF
    -D USE_FFMPEG=OFF
    -D USE_FIELD3D=OFF
    -D USE_GIF=OFF
    -D USE_HDF5=OFF
    -D USE_LIBHEIF=OFF
    -D USE_OPENCV=OFF
    -D USE_OPENJPEG=OFF
    -D USE_OPENVDB=OFF
    -D USE_PTEX=OFF
    -D USE_QT=OFF
    -D USE_WEBP=OFF
    -D STOP_ON_WARNING=OFF
    -D VERBOSE=ON
```

**Relevant output for the broken build (against the split-release of IlmBase/OpenEXR):**

Find information for OpenEXR:
```
-- Found OpenEXR: /luma/soft/rez/packages/ilmbase/2.5.5/platform-linux/gcc-4.8.3/include  
-- Found OpenEXR 2.5.5 
--     OPENEXR_INCLUDES = /luma/soft/rez/packages/openexr/2.5.5/platform-linux/gcc-4.8.3/include
--     OPENEXR_INCLUDE_DIR = /luma/soft/rez/packages/openexr/2.5.5/platform-linux/gcc-4.8.3/include
--     OPENEXR_LIBRARIES = /luma/soft/rez/packages/openexr/2.5.5/platform-linux/gcc-4.8.3/lib64/libIlmImf-2_5.a;/luma/soft/rez/packages/ilmbase/2.5.5/platform-linux/gcc-4.8.3/lib64/libImath-2_5.a;/luma/soft/rez/packages/ilmbase/2.5.5/platform-linux/gcc-4.8.3/lib64/libIex-2_5.a;/luma/soft/rez/packages/ilmbase/2.5.5/platform-linux/gcc-4.8.3/lib64/libHalf-2_5.a;/luma/soft/rez/packages/ilmbase/2.5.5/platform-linux/gcc-4.8.3/lib64/libIlmThread-2_5.a;-lpthread;/usr/lib64/libz.so
--     IMATH_INCLUDES = /luma/soft/rez/packages/ilmbase/2.5.5/platform-linux/gcc-4.8.3/include
```

First point of failure:
```
[  0%] Building CXX object src/libutil/CMakeFiles/OpenImageIO_Util.dir/argparse.cpp.o
cd /mnt/bonus/dev/rez-build/openimageio/dev/build/platform-linux/gcc-4.8.3/boost==1.61.0/python-2.7/!nuke/src/libutil && ccache /usr/bin/g++  -DOpenImageIO_EXPORTS -DOpenImageIO_Util_EXPORTS -DUSE_BOOST_REGEX -DUSE_FREETYPE=1 -DUSE_JPEG_TURBO=1 -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -I/mnt/bonus/dev/rez-build/openimageio/dev/build/platform-linux/gcc-4.8.3/boost==1.61.0/python-2.7/!nuke/include/OpenImageIO -I/mnt/bonus/dev/rez-build/openimageio/dev/build/platform-linux/gcc-4.8.3/boost==1.61.0/python-2.7/!nuke/include -I/mnt/bonus/dev/rez-build/openimageio/dev/build/platform-linux/gcc-4.8.3/boost==1.61.0/python-2.7/!nuke/src/include -I/mnt/bonus/dev/rez-build/openimageio/dev/src/src/include -I/luma/soft/rez/packages/fmt/6.1.2/platform-linux/include -I/luma/dev/ruschn/rez-release/openimageio/2.2.13.0/platform-linux/gcc-4.8.3/boost==1.61.0/python-2.7/!nuke/include -isystem /luma/soft/rez/packages/ilmbase/2.5.5/platform-linux/gcc-4.8.3/include -isystem /luma/soft/rez/packages/boost/1.61.0/platform-linux/gcc-4.8.3/python-2.7/include -isystem /luma/soft/rez/packages/ilmbase/2.5.5/platform-linux/gcc-4.8.3/include/OpenEXR  -O3 -DNDEBUG -fPIC   -Wall -Wno-unused-local-typedefs -Wno-unused-result -fno-math-errno -msse4.2 -std=c++11 -o CMakeFiles/OpenImageIO_Util.dir/argparse.cpp.o -c /mnt/bonus/dev/rez-build/openimageio/dev/src/src/libutil/argparse.cpp
In file included from /mnt/bonus/dev/rez-build/openimageio/dev/src/src/include/OpenImageIO/fmath.h:40:0,
                 from /mnt/bonus/dev/rez-build/openimageio/dev/src/src/include/OpenImageIO/hash.h:31,
                 from /mnt/bonus/dev/rez-build/openimageio/dev/src/src/include/OpenImageIO/strutil.h:24,
                 from /mnt/bonus/dev/rez-build/openimageio/dev/src/src/include/OpenImageIO/typedesc.h:29,
                 from /mnt/bonus/dev/rez-build/openimageio/dev/src/src/include/OpenImageIO/attrdelegate.h:13,
                 from /mnt/bonus/dev/rez-build/openimageio/dev/src/src/include/OpenImageIO/paramlist.h:17,
                 from /mnt/bonus/dev/rez-build/openimageio/dev/src/src/include/OpenImageIO/argparse.h:20,
                 from /mnt/bonus/dev/rez-build/openimageio/dev/src/src/libutil/argparse.cpp:16:
/mnt/bonus/dev/rez-build/openimageio/dev/src/src/include/OpenImageIO/Imath.h:13:35: fatal error: OpenEXR/OpenEXRConfig.h: No such file or directory
 #include <OpenEXR/OpenEXRConfig.h>
```

The basic issue is that `OpenEXR/OpenEXRConfig.h` is installed as part of OpenEXR (not IlmBase), but the compile command is only including the IlmBase include dirs.

**Relevant output for the working build (against combined IlmBase/OpenEXR install):**

Find information for OpenEXR:
```
-- Found OpenEXR 2.5.5 
--     OPENEXR_INCLUDES = /luma/dev/ruschn/rez-release/openexr/2.5.5/platform-linux/gcc-4.8.3/include
--     OPENEXR_INCLUDE_DIR = /luma/dev/ruschn/rez-release/openexr/2.5.5/platform-linux/gcc-4.8.3/include
--     OPENEXR_LIBRARIES = /luma/dev/ruschn/rez-release/openexr/2.5.5/platform-linux/gcc-4.8.3/lib64/libIlmImf-2_5.a;/luma/dev/ruschn/rez-release/openexr/2.5.5/platform-linux/gcc-4.8.3/lib64/libImath-2_5.a;/luma/dev/ruschn/rez-release/openexr/2.5.5/platform-linux/gcc-4.8.3/lib64/libIex-2_5.a;/luma/dev/ruschn/rez-release/openexr/2.5.5/platform-linux/gcc-4.8.3/lib64/libHalf-2_5.a;/luma/dev/ruschn/rez-release/openexr/2.5.5/platform-linux/gcc-4.8.3/lib64/libIlmThread-2_5.a;-lpthread;/usr/lib64/libz.so
--     IMATH_INCLUDES = /luma/dev/ruschn/rez-release/openexr/2.5.5/platform-linux/gcc-4.8.3/include
```

Working compile command for same object as above:
```
[  0%] Building CXX object src/libutil/CMakeFiles/OpenImageIO_Util.dir/argparse.cpp.o
cd /mnt/bonus/dev/rez-build/openimageio/dev/build/platform-linux/gcc-4.8.3/boost==1.61.0/python-2.7/!nuke/src/libutil && ccache /usr/bin/g++  -DOpenImageIO_EXPORTS -DOpenImageIO_Util_EXPORTS -DUSE_BOOST_REGEX -DUSE_FREETYPE=1 -DUSE_JPEG_TURBO=1 -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -I/mnt/bonus/dev/rez-build/openimageio/dev/build/platform-linux/gcc-4.8.3/boost==1.61.0/python-2.7/!nuke/include/OpenImageIO -I/mnt/bonus/dev/rez-build/openimageio/dev/build/platform-linux/gcc-4.8.3/boost==1.61.0/python-2.7/!nuke/include -I/mnt/bonus/dev/rez-build/openimageio/dev/build/platform-linux/gcc-4.8.3/boost==1.61.0/python-2.7/!nuke/src/include -I/mnt/bonus/dev/rez-build/openimageio/dev/src/src/include -I/luma/soft/rez/packages/fmt/6.1.2/platform-linux/include -I/luma/dev/ruschn/rez-release/openimageio/2.2.13.0/platform-linux/gcc-4.8.3/boost==1.61.0/python-2.7/!nuke/include -isystem /luma/dev/ruschn/rez-release/openexr/2.5.5/platform-linux/gcc-4.8.3/include -isystem /luma/soft/rez/packages/boost/1.61.0/platform-linux/gcc-4.8.3/python-2.7/include -isystem /luma/dev/ruschn/rez-release/openexr/2.5.5/platform-linux/gcc-4.8.3/include/OpenEXR  -O3 -DNDEBUG -fPIC   -Wall -Wno-unused-local-typedefs -Wno-unused-result -fno-math-errno -msse4.2 -std=c++11 -o CMakeFiles/OpenImageIO_Util.dir/argparse.cpp.o -c /mnt/bonus/dev/rez-build/openimageio/dev/src/src/libutil/argparse.cpp
```

**Comment by lgritz** [comment]
I see what's happening.

Unfortunately, OpenEXR 2.5 had the unfortunate habit of exporting config files that had the include path set to ".../prefix/include/OpenEXR" instead of ".../prefix/include". OpenEXR/Imath 3.0 fix this to conform to the more usual convention.

I'll try to fix up on the OIIO side to deal with this more gracefully.

**Comment by lgritz** [comment]
Wait, maybe I'm wrong about the cause. Poking around.

**Comment by lgritz** [comment]
I see, split tree is the crux.

**Comment by lgritz** [comment]
How do you engineer the split install? When I build OpenEXR 2.5 with cmake, there is only one CMAKE_INSTALL_PREFIX.

**Comment by nrusch** [comment]
The pre-3.0 codebase includes IlmBase and OpenEXR as separate CMake projects, and then the root "meta-project" just builds all of the child projects at once using a single set of config variables. Thus, if you want to build and install IlmBase by itself, for example, you can just point CMake at [the IlmBase subdirectory](https://github.com/AcademySoftwareFoundation/openexr/tree/RB-2.5/IlmBase).

In our case, we just have separate rez packages and build scripts for each one, which again ties back to the historical separation between the two. I am considering switching to a unified build, but given that there are clean lines that let you build and install each one separately, it seems like there could still be merit in OIIO supporting that kind of release pattern (at least until 3.0 becomes the minimum required version).

**Comment by lgritz** [comment]
This breakage is not intentional. I'll try to look at this issue today.

We have a complicated code path for finding the elements of OpenEXR/Ilmbase/Imath, to account for an especially wide range of versions. We must have botched things.

Do you know if you have exported *Config.cmake files for your separated installations? Or is this in the no-config codepath?

**Comment by nrusch** [comment]
Yes, we are installing the *Config.cmake files for both, and adding both the `<prefix>` and `<prefix>/lib64/cmake/[IlmBase|OpenEXR]` directories to `CMAKE_PREFIX_PATH` in our build environments.

In other words, when OIIO is building, these directories are all on `CMAKE_PREFIX_PATH`:
- `<IlmBase_PREFIX>`
- `<IlmBase_PREFIX>/lib64/cmake/IlmBase`
- `<OpenEXR_PREFIX>`
- `<OpenEXR_PREFIX>/lib64/cmake/OpenEXR`

**Comment by lgritz** [comment]
Do you think you could please tar/zip up and email me (lg@larrygritz.com) the full set of the relevant config files in {IlmBase,OpenEXR}/lib64/cmake just so I can be extra sure I understand what's being set? Thanks.

**Comment by nrusch** [comment]
Yeah, I'll send them over tomorrow when I'm back at my desk.

**Comment by nrusch** [comment]
OK, just sent them over.

**Comment by lgritz** [comment]
Proposed fix in #2935. It would be very helpful if you were able to apply that patch on your end and confirm whether or not it solves the problem with your particular setup.
